File: src/main/ai/tools/builtin/session_status.ts
/**
 * session_status — 当前会话状态查询
 *
 * 让 Agent 自我感知当前会话的运行状态：
 *   - 会话 ID、运行时长
 *   - contexts/ 快照数量
 *   - 最近一次 LLM 调用的 token 用量、模型、耗时
 *
 * 数据来源：workspace/contexts/*.json
 *
 * 分类：Observability | 风险：低（只读）
 */

import fs from 'node:fs'
import path from 'node:path'
import { z } from 'zod'
import type { ToolDefinition, ToolStreamUpdate, ToolResult } from '../types'
import { ToolCategory } from '../types'

export const sessionStatusTool: ToolDefinition = {
  name: 'session_status',
  description:
    'View current session status: session ID, snapshot count, ' +
    'last LLM call info (model, duration, token usage). ' +
    'Use this for self-monitoring and cost awareness.',
  category: ToolCategory.Observability,
  needUserConfirm: false,
  parameters: z.object({}),

  execute: async function* (
    _params: Record<string, unknown>,
    _signal?: AbortSignal,
    context?: { workspaceRoot: string; sessionId?: string }
  ): AsyncGenerator<ToolStreamUpdate, ToolResult, unknown> {
    yield { type: 'progress' as const, content: '[session_status] querying...' }

    const workspace = context?.workspaceRoot
    if (!workspace) {
      return { success: false, llmContent: 'Error: workspace not available in context.' }
    }

    const contextsDir = path.join(workspace, 'contexts')

    if (!fs.existsSync(contextsDir)) {
      return {
        success: true,
        llmContent: `Session: ${context.sessionId || 'unknown'}\nSnapshots: 0\nNo LLM calls recorded yet.`
      }
    }

    // 列出所有 context JSON 文件
    const files = fs
      .readdirSync(contextsDir)
      .filter((f) => f.endsWith('.json'))
      .sort()

    const snapshotCount = files.length

    // 读取最近一次快照
    let lastInfo = 'No snapshots available.'
    if (files.length > 0) {
      try {
        const lastFile = files[files.length - 1]
        const raw = fs.readFileSync(path.join(contextsDir, lastFile), 'utf-8')
        const snap = JSON.parse(raw)

        const parts: string[] = []
        parts.push(`Last call: ${snap.timestamp || lastFile}`)
        if (snap.config?.model) parts.push(`Model: ${snap.config.model}`)
        if (snap.duration !== undefined) parts.push(`Duration: ${snap.duration}ms`)
        if (snap.toolCalls?.length) parts.push(`Tool calls: ${snap.toolCalls.length}`)
        if (snap.error) parts.push(`Error: ${snap.error}`)
        lastInfo = parts.join('\n')
      } catch {
        lastInfo = 'Failed to parse last snapshot.'
      }
    }

    const output = [
      `Session: ${context.sessionId || 'unknown'}`,
      `Snapshots: ${snapshotCount}`,
      '',
      lastInfo
    ].join('\n')

    return { success: true, llmContent: output }
  }
}

File: skills/observability/scripts/session-status.py
#!/usr/bin/env python3
"""
Session Status Script - 会话状态查询

用途：
    查看当前会话的运行状态：会话 ID、快照数量、最后一次 LLM 调用信息

使用方式：
    python session-status.py

输出格式（文本）：
    Session: abc123-def456-...
    Snapshots: 15
    
    Last call: 2024-02-23T09:30:42.123Z
    Model: dashscope/qwen3.5-plus
    Duration: 2345ms
    Tool calls: 3

作者：Coobee AI Team
"""

import json
import os
import sys
from pathlib import Path


def get_session_status():
    """获取会话状态"""
    try:
        # 从环境变量获取工作空间目录
        workspace = os.environ.get("COOBEE_WORKSPACE")
        
        if not workspace:
            # 降级方案：使用当前工作目录
            workspace = str(Path.cwd())
        
        session_id = os.environ.get("COOBEE_SESSION_ID", "unknown")
        
        contexts_dir = Path(workspace) / "contexts"
        
        # 检查目录是否存在
        if not contexts_dir.exists():
            print(f"Session: {session_id}")
            print("Snapshots: 0")
            print("\nNo LLM calls recorded yet.")
            return
        
        # 列出所有 context JSON 文件
        files = sorted([f for f in contexts_dir.iterdir() if f.suffix == '.json'])
        snapshot_count = len(files)
        
        print(f"Session: {session_id}")
        print(f"Snapshots: {snapshot_count}")
        
        # 读取最近一次快照
        if files:
            try:
                last_file = files[-1]
                with open(last_file, 'r', encoding='utf-8') as f:
                    snap = json.load(f)
                
                print()  # 空行
                
                # 提取信息
                if 'timestamp' in snap:
                    print(f"Last call: {snap['timestamp']}")
                else:
                    print(f"Last call: {last_file.name}")
                
                if 'config' in snap and 'model' in snap['config']:
                    print(f"Model: {snap['config']['model']}")
                
                if 'duration' in snap:
                    print(f"Duration: {snap['duration']}ms")
                
                if 'toolCalls' in snap and snap['toolCalls']:
                    print(f"Tool calls: {len(snap['toolCalls'])}")
                
                if 'error' in snap:
                    print(f"Error: {snap['error']}")
                    
            except Exception as e:
                print(f"\nFailed to parse last snapshot: {e}")
        else:
            print("\nNo snapshots available.")
    
    except Exception as e:
        error_msg = {
            "error": "Failed to get session status",
            "message": str(e)
        }
        print(json.dumps(error_msg, ensure_ascii=False), file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    get_session_status()
